src/middlewared/middlewared/plugins/zfs.py

```diff
@@ -326,6 +327,141 @@ def promote(self, name):
+
+    @accepts(Dict(
+        'snapshot-task',
+        Str('task_name'),
+        Str('task_filesystem'),
+        Bool('task_recursive'),
+        Int('task_ret_count'),
+        Str('task_ret_unit'),
+        Str('task_begin'),
+        Str('task_end'),
+        Int('task_interval'),
+        Str('task_repeat_unit'),
+        Str('task_byweekday'),
+        Bool('task_enabled'),
+        Str('task_last_run'),
+        Int('vmware_snap_task_id'),
+        register=True,
+    ))
+    async def do_create(self, data):
+        return await self.middleware.call(
+            'datastore.insert',
+            'storage.task',
+            data,
+        )
+
+    @accepts(Int('id'), Ref('snapshot-task'))
+    async def do_update(self, id, data):
+        return await self.middleware.call(
+            'datastore.update',
+            'storage.task',
+            id,
+            data,
+        )
+
+    @accepts(Int('id'))
+    async def do_delete(self, id):
+        return await self.middleware.call(
+            'datastore.delete',
+            'storage.task',
+            id
+        )
+
+    @job(lock='autosnap')
+    @periodic(60)
+    def autosnap(self, job):
+        snap_tasks = self.middleware.call_sync('zfs.snapshot_task.query', [('task_enabled', '=', True)])
+        zfs = libzfs.ZFS()
+        tasks_to_execute = []
+
+        date_now = datetime.datetime.now()
+        curtime = datetime.time(date_now.hour, date_now.minute)
+
+        for task in snap_tasks:
+            if task['task_last_run']:
+                snaptask_time_delta = task['task_last_run'] + datetime.timedelta(minutes=task['task_interval'])
+            else:
+                snaptask_time_delta = date_now
+
+            if curtime < task['task_begin'] or curtime > task['task_end'] or date_now < snaptask_time_delta or str(date_now.weekday()+1) not in task['task_byweekday']:
+                continue
+
+            try:
+                zfs.get_dataset(task['task_filesystem'])
+            except libzfs.ZFSException as err:
+                self.logger.error(f'{err} - Periodic snapshot task omitted.')
+            tasks_to_execute.append(task)
+
+        for task in tasks_to_execute:
+            task_name = task['task_name'] if task['task_name'] else f"task_{task['id']}"
+            self.middleware.call_sync(
+                'zfs.snapshot.do_create',
+                {'dataset': task['task_filesystem'],
+                'recursive': task['task_recursive'],
+                'name': f'autosnap_new_{task_name}_{date_now.strftime("%Y%m%d.%H%M")}',
+                'properties': {'autosnap:name': task_name, 'autosnap:retention': f'{task["task_ret_count"]}:{task["task_ret_unit"]}'}}
+            )
+
+            self.middleware.call_sync(
+                'zfs.snapshot_task.do_update',
+                task['id'],
+                {'task_last_run': f'{date_now.year}-{date_now.month}-{date_now.day} {date_now.hour}:{date_now.minute}'}
```

Comment:
The reason we have avoided storing last_run on database is because it can wear usb flash disks. Imagine you have a task running every minute, it means it will write to flash drive every minute (or close).